### PR TITLE
fix: mark sender in visited bloom filter to prevent GET request cycles

### PR DESCRIPTION
**Pull request by sanity**
## Problem

When processing a forwarded GET request, the code was only marking the current peer (`this_peer`) in the visited bloom filter, but **not the sender** (`source_addr`). This allowed requests to cycle back to nodes that had already seen the request.

**User-visible impact:** Nodes with limited connections (e.g., only connected to gateway) would see GET requests fail with "no peers to forward to and no upstream" error when requesting contracts that don't exist in the reachable network.

### Bug scenario (before fix)

1. Node A (with single connection to gateway) sends GET for non-local contract
2. Gateway marks itself in visited, forwards to other peers  
3. Other peers mark themselves, but **NOT** the sender
4. Eventually routing sends request back to Node A (not in visited filter)
5. Node A can't forward (only peer is gateway, already in filter)
6. Request fails with "no peers to forward to and no upstream"

This was discovered through user logs where a peer with only one connection to the gateway was unable to fetch contracts.

### Why CI didn't catch this

The bug only manifests when:
- A node has very limited connections (1 or very few)
- The contract doesn't exist anywhere in the reachable network
- Routing decides to send the request back to the originator

CI tests typically have well-connected multi-peer networks where even if a request cycles back, nodes have alternative peers to try. The single-connection edge case combined with a non-existent contract wasn't covered.

## Approach

Mark `source_addr` (the sender) alongside `this_peer` when processing received GET requests. This matches the behavior already present in `subscribe.rs` at line 517-518 where `requester_addr` is marked.

The key insight is that we don't want the **same request** to visit **any peer twice** - not just the current node, but also any node that previously forwarded the request to us.

## Testing

- Added unit tests in `visited_peers.rs` documenting the expected behavior:
  - `test_forwarding_must_mark_both_this_peer_and_sender` - verifies both addresses must be marked
  - `test_request_cycle_prevention` - simulates complete routing chain and verifies all nodes are blocked

### CI coverage gap

For broader CI coverage of limited-connectivity scenarios, consider adding an integration test that:
1. Creates a peer with only 1 connection (to gateway)
2. Makes GET request for contract that doesn't exist in the network
3. Verifies the request fails gracefully without cycling back to originator

This would require test infrastructure changes to simulate limited connectivity, which could be a follow-up issue.

[AI-assisted - Claude]

### CODE REVIEW

**Comment by sanity** [comment]
@claude Please skeptically review this PR

**Comment by sanity** [comment]
## Responses to Review Feedback

### Claude's Skeptical Review - `try_forward_or_return` concern

**Not an issue.** The `visited` filter passed to `try_forward_or_return` (line 1003: `new_visited`) already has the sender marked from the fix at lines 874-876. The function operates on a clone that already contains both `this_peer` and `source_addr`. The redundant `mark_visited(this_peer)` inside the function is harmless for a bloom filter.

The flow is:
1. Lines 867-876: Create `new_visited`, mark both `this_peer` and `source_addr` ✓
2. Line 1003: Pass this `new_visited` to `try_forward_or_return`
3. Lines 1796-1799: Function marks `this_peer` again (no-op, already marked)

### Copilot Inline Comments

1. **Confusing test comment** (line 463) - Fixed in c3e467cc. Rewrote the test to clearly show the complete marking sequence at each hop.

2. **Test not validating the fix** - This is intentional. The tests are *documentation tests* that show the expected behavior pattern, not integration tests of the actual code path. The follow-up issue #2417 tracks adding a proper integration test for the full scenario.

[AI-assisted - Claude]

File: crates/core/src/operations/visited_peers.rs [hunk 1]
```diff
@@ -394,4 +394,100 @@ mod tests {
...
+
+        // Simulate the routing chain:
+        // Originator (A) -> Gateway (G) -> Peer1 (P1) -> Peer2 (P2)
+        let originator: SocketAddr = "10.0.0.1:8000".parse().unwrap();
+        let gateway: SocketAddr = "10.0.0.2:8000".parse().unwrap();
+        let peer1: SocketAddr = "10.0.0.3:8000".parse().unwrap();
+        let peer2: SocketAddr = "10.0.0.4:8000".parse().unwrap();
+
+        // Step 1: Originator should mark themselves BEFORE sending (not tested here, but expected)
+        // For now, originator is NOT in visited - this is the gap that caused the bug
```
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/src/operations/visited_peers.rs]: Fixed in c3e467cc - rewrote the test comments to clearly show the bug context and the complete marking sequence at each hop.

[AI-assisted - Claude]

File: crates/core/src/operations/visited_peers.rs [hunk 2]
```diff
@@ -394,4 +394,100 @@ mod tests {
...
+        visited.mark_visited(gateway);
+        visited.mark_visited(originator); // KEY: Mark sender too!
+
+        // Step 3: Peer1 receives from gateway, marks peer1 AND gateway
+        visited.mark_visited(peer1);
+        // gateway already marked, but would be marked again in real code
+
+        // Step 4: Peer2 receives from peer1, marks peer2 AND peer1
+        visited.mark_visited(peer2);
+        // peer1 already marked
```
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/src/operations/visited_peers.rs]: This is intentional - these are *documentation tests* that demonstrate the expected behavior pattern, not integration tests of the actual code path. They verify that the bloom filter works correctly when used as intended.

A proper integration test for the full scenario (limited-connectivity node, request routing, cycle prevention) is tracked in follow-up issue #2417.

[AI-assisted - Claude]
